### PR TITLE
INFRA-1038 Clever logger for Coffee

### PR DESCRIPTION
**Pull request by peternga**
Port from https://github.com/Clever/kayvee-go/pull/9

### CODE REVIEW

File: lib/clog/clog.coffee [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+class Logger
+  constructor: (source, logLvl=null, @formatter=kv.format, output=console.error) ->
+    if !logLvl?
+      strLogLvl = process.env.LOG_LEVEL_CONFIG
+      if !strLogLvl?
+        logLvl = LOG_LEVELS.Debug
+      else
+        vals = (val for key, val of LOG_LEVELS)
+        logLvl = if (strLogLvl in vals) then strLogLvl else LOG_LEVELS.Debug
```
Comment (mohit) [inline lib/clog/clog.coffee]: we should log the log level we end up using :)

File: lib/clog/clog.coffee [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    @logLvl = logLvl
+    @globals = {}
+    @globals["source"] = source
+    @logWriter = output
+
+  setConfig: (source, logLvl, formatter, output) ->
+    @source = source
+    @logLvl = logLvl
```
Comment (mohit) [inline lib/clog/clog.coffee]: should we not verify the logLvl here just as we do in the constructor. Maybe one way to do this is to move out the validation logic into an internal method. 

then we could have in the `constructor`:

```coffeescript
strLogLvl = process.env.LOG_LEVEL_COFIG or LOG_LEVELS.Debug
logLvl = _validateLogLvl(strLogLvl)
// log the logLvl we are using
```

and in `setConfig`
```
@logLvl = _validateLogLvl(logLvl)
```

File: lib/clog/clog.coffee (hunk 1 above)
Comment (peternga) [inline lib/clog/clog.coffee]: The log level gets used when we're actually logging. See logWithLevel

File: lib/clog/clog.coffee (hunk 2 above)
Comment (peternga) [inline lib/clog/clog.coffee]: Agreed. Let me address this

File: lib/clog/clog.coffee [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  _validateLogLvl: (logLvl) ->
+    if !logLvl?
+      return LOG_LEVELS.Debug
+    else
+      for own key, value of LOG_LEVELS
+        if logLvl.toLowerCase() == value
+          return value
+
+  setLogLevel: (logLvl) ->
+    @logLvl = logLvl
```
Comment (mohit) [inline lib/clog/clog.coffee]:
```coffeescript
@logLvl = validatateLogLvl(logLvl)
```

File: lib/clog/clog.coffee [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    @globals["source"] = source
+    @logWriter = output
+
+  setConfig: (source, logLvl, formatter, output) ->
+    @global["source"] = source
+    @logLvl = @_validateLogLvl(logLvl)
+    @formatter = formatter
+    @logWriter = output
+
+  _validateLogLvl: (logLvl) ->
```
Comment (mohit) [inline lib/clog/clog.coffee]: maybe add a test for this?

File: lib/clog/clog.coffee [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+LOG_LEVEL_ENUM = {
+  "debug":    0,
+  "info":     1,
+  "warning":  2,
+  "error":    3,
+  "critical": 4,
+}
+
+class Logger
+  constructor: (source, logLvl=null, @formatter=kv.format, output=console.error) ->
```
Comment (mohit) [inline lib/clog/clog.coffee]: the tests are lacking use of the constructor.

**Comment by mohit** [comment]
LGTM! :+1: :fireworks: this is so exciting!

Some nit's about the testing the constructor and validation, but otherwise looks awesome?

**Comment by peternga** [comment]
Minor changes according to reviews and comments in other languages of the library to maintain consistency

File: test/logger_test.coffee [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+      logLvl = logObj._validateLogLvl "Debug"
+      assert.equal logLvl, logger.Debug
+
+      logLvl = logObj._validateLogLvl "info"
+      assert.equal logLvl, logger.Info
+      logLvl = logObj._validateLogLvl "Info"
+      assert.equal logLvl, logger.Info
+
+      logLvl = logObj._validateLogLvl "sometest"
+      assert.equal logLvl, logger.Debug
```
Comment (nathanleiby) [inline test/logger_test.coffee]: To make these test cases more explicit/readable, I might consider a pattern like:

```
describe '.validatelog', ->
   it 'is case-insensitive in log level name', ->
      logLvl = logObj._validateLogLvl "debug"
      assert.equal logLvl, logger.Debug
      logLvl = logObj._validateLogLvl "Debug"
      assert.equal logLvl, logger.Debug

  it 'sets non-default log levels', ->
      logLvl = logObj._validateLogLvl "info"
      assert.equal logLvl, logger.Info
      # TODO: for each possible log level ...
      logLvl = logObj._validateLogLvl "critical"
      assert.equal logLvl, logger.Critical

  it 'sets level to Debug, if given an invalid log level', ->
      logLvl = logObj._validateLogLvl "sometest"
      assert.equal logLvl, logger.Debug
```

File: test/logger_test.coffee [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+  describe '.invalidlog', ->
+    it 'invalidlog', ->
+      # Invalid log levels will default to debug
+      logObj.setLogLevel "debu"
```
Comment (nathanleiby) [inline test/logger_test.coffee]: typo: debu/debug

Comment (nathanleiby) [inline test/logger_test.coffee]: Derp, nvm :) thought this was the testcase below!

File: test/logger_test.coffee [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+      formatter = (data) -> data["level"] + "." + data["source"] + "." + data["title"]
+      logObj = new logger("logger-test", logger.Info, formatter, outputFunc)
+      logObj.debug "testlogdebug"
+      expected = ""
+      assert.equal sample, expected
+
+      logObj.info "testloginfo"
+      expected = logger.Info + ".logger-test.testloginfo"
+      assert.equal sample, expected
+  describe '.validatelog', ->
```
Comment (nathanleiby) [inline test/logger_test.coffee]: s/validatelog/_validateLogLvl?

File: test/logger_test.coffee (hunk 6 above)
Comment (nathanleiby) [inline test/logger_test.coffee]: This comment applies throughout. e.g.

similarly, describing in a sentence would make things more clear.

```
describe diffoutput 
  it does something
```

**Comment by nathanleiby** [comment]
LGTM, just added some style suggestions

**Pull request merged**